examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,624 @@
+"""
+================
+Metadata Routing
+================
+
+.. currentmodule:: sklearn
+
+This document shows how you can use the metadata routing mechanism in
+scikit-learn to route metadata through meta-estimators to the estimators using
+them. To better understand the rest of the document, we need to introduce two
+concepts: routers and consumers. A router is an object, in most cases a
+meta-estimator, which routes given data and metadata to other objects and
+estimators. A consumer, on the other hand, is an object which accepts and uses
+a certain given metadata. For instance, an estimator taking into account
+``sample_weight`` is a consumer of ``sample_weight``. It is possible for an
+object to be both a router and a consumer. For instance, a meta-estimator may
+take into account ``sample_weight`` in certain calculations, but it may also
+route it to the underlying estimator.
+
+First a few imports and some random data for the rest of the script.
+"""
+# %%
+
+import numpy as np
+import warnings
+from sklearn.base import BaseEstimator
+from sklearn.base import ClassifierMixin
+from sklearn.base import RegressorMixin
+from sklearn.base import MetaEstimatorMixin
+from sklearn.base import TransformerMixin
+from sklearn.base import clone
+from sklearn.utils.metadata_requests import RequestType
+from sklearn.utils.metadata_requests import metadata_request_factory
+from sklearn.utils.metadata_requests import MetadataRouter
+from sklearn.utils.validation import check_is_fitted
+from sklearn.linear_model import LinearRegression
+
+N, M = 100, 4
+X = np.random.rand(N, M)
+y = np.random.randint(0, 2, size=N)
+my_groups = np.random.randint(0, 10, size=N)
+my_weights = np.random.rand(N)
+my_other_weights = np.random.rand(N)
+
+# %%
+# Estimators
+# ----------
+# Here we demonstrate how an estimator can expose the required API to support
+# metadata routing as a consumer. Imagine a simple classifier accepting ``foo``
+# as a metadata on its ``fit`` and ``bar`` in its ``predict`` method. We add
+# two constructor arguments to helps us check whether an expected metadata is
+# given or not. This is a minimal scikit-learn compatible classifier:
+
+
+class ExampleClassifier(ClassifierMixin, BaseEstimator):
+    def __init__(self, foo_is_none=True, bar_is_none=True):
```

Comment:
I think that in the final version of the documentation, I would find it easier to have names of known scikit-learn parameters such as `sample_weight`.